### PR TITLE
Easier plugin development

### PR DESCRIPTION
**Pull request by l0drex**
This pull request comes with a new module containing abstract base classes for plugins. This should make the development of plugins way easier, remove redundancy and allow iterative working with those plugins as they all share the same interface.

### ToDos

- [x] Create such a module
- [x] Change all existing plugins so that they use this plugin
- [x] Work with these new plugins in yin_yang, config and gui iteratively to reduce redundancy
- [x] Add tests
- [x] Provide good documentation
- [x] Provide backwards compatibility
- [x] Test plugins set_theme() method (see comment below)

### CODE REVIEW

**Comment by l0drex** [comment]
_I'm pasting the documentation here since I don't know yet if I should put them into contributing.md or the wiki. Also, it should provide a good overview on the changes I made here._

---

## Plugins

Plugins provide a way to directly change the theme by either running a command or
change a value in a config file.

Each plugin is a class in `src/plugins` and extends the class `Plugin` in `_plugin.py`.
Write your code to set any theme by overwriting the method `set_theme()`. An easy example would be `kvantum.py`.


### Default themes

If you want to provide default values for light and dark themes,
you can set the class attributes `theme_light` and `theme_dark`.

If you have no such default values, there will be a warning in the terminal, but in production this is just ignored.


### GUI

By default, each plugin has its own section in the tab plugins,
with one simple text input to set the desired theme.

If you want to provide a combo box instead, you can simply implement the property method `available_themes`.


### Testing

Testing plugins currently only really checks that none of the enabled plugins `set_theme` functions
result in an error, plus some setup stuff.
To test your plugin you have to:
1. Enable it (via UI or the config file stored in `~/.config/yin-yang/yin-yang.json`).
2. Change the `False` in `tests/test_plugins.py: test_set_theme_works()` to `True`
2. Run `tests/test_plugins.py`.

Alternatives are:
- use [assertions](https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement):
  ```python
  assert statement, 'Warning if statement is false'
  ```
- use the python console

**Comment by l0drex** [comment]
The config page now looks a bit different, as in of my previous pull requests (#68). I could also move the checkbox next to the title, which would hide that big gap between checkbox and dropdown or buttons.

grafik

**Comment by l0drex** [comment]
### Tested plugins

- [x] KDE
- [x] Kvantum

**Comment by oskarsh** [comment]
Hey @l0drex, your PR looks really good so far. I would move the checkboxes back next to the title to avoid that awkward space and have it more uniform. I will give it a proper review in the following days. Thank you for your work!

**Comment by l0drex** [comment]
> I would move the checkboxes back next to the title to avoid that awkward space and have it more uniform.

Done. I also updated some test methods.

**Comment by l0drex** [comment]
@oskarsh by the way, could you please look at #81 first? It's a smaller pull request regarding the communication between this application and the Firefox plugin. If #81 gets merged, I can update the plugin to support the newer messages.

File: src/config.py [hunk 1]
```diff
@@ -79,39 +79,79 @@ def set_sun_time():
...
+else:
+    # TODO add default gtk themes for non-kde systems
+    config["gtkLightTheme"] = ""
+    config["gtkDarkTheme"] = ""
+
+config["atomEnabled"] = False
+config["atomLightTheme"] = "one-light"
+config["atomDarkTheme"] = "one-dark"
+
+# TODO add default wallpapers
```
Comment (oskarsh) [inline src/config.py:111]: totally fine to provide no default wallpaper

File: src/plugins/__init__.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+from src.plugins import system, gtk, kvantum, wallpaper, firefox, vscode, atom
+
+# NOTE initialize your plugin over here:
+# The order in the list specifies the order in the config gui
+from src.plugins._plugin import Plugin
+
```
Comment (oskarsh) [inline src/plugins/__init__.py:6]: nice job

**Comment by oskarsh** [comment]
once you updated the plugin to support the newer messages I am fine with merging this! Really nice work thank you!

**Comment by l0drex** [comment]
@oskarsh Communication is now merged and there are no conflicts anymore.

**Comment by l0drex** [comment]
Don't forget to update the wiki ;)

**Pull request merged**